src/lib.rs

```diff
@@ -405,6 +408,19 @@ fn get_num_cpus() -> usize {
     unsafe { hermit_abi::get_processor_count() }
 }
 
+#[cfg(feature = "wasm-bindgen")]
```

Comment:
These should be gated by the arch as well.

Reply:
It's already gated by arch in Cargo.toml. There is no way to enable this feature on the wrong arch.

I don't think we need to explicitly exclude wasi/emscripten/etc. If user explicitly opted into wasm-bindgen, we can assume they know what they're doing, otherwise it won't compile anyway.

In fact, I'm hoping for wasm-bindgen to support WASI target someday, so this way code is more future-proof.

Reply:
There are no target specific features (until resolver 2 lands eventually). This code fails to compile if the feature is active and you compile for a different target, which is not super unlikely. The gate in the Cargo.toml does nothing if the code itself can't find the crate.

Reply:
Apparently the resolver can only be set by the top level crate. This means that dependencies like num-cpus will never be able to assume that there are target specific features.

Reply:
> This code fails to compile if the feature is active and you compile for a different target, which is not super unlikely.

That's kinda the goal anyway - to fail to compile on wrong target/feature combination, so we still know that `wasm-bindgen` feature will only compile on wasm32 architecture and don't need to spell out that explicitly in the code.

Reply:
It's also important to note that in this case it's not just a feature, but an optional dependency, which won't be even fetched for non-wasm32 architecture by Cargo.

Reply:
The point is that by cargo's current design (and unfortunately kinda future design too) there is no such thing as a wrong target and feature combination. Cargo CAN NOT (in general) support this. The wasm-bindgen feature might be active for other targets because cargo's feature resolver has no notion of targets. It assigns all features to all targets. So with this you may not be able to compile for Linux anymore.

Reply:
I'm confused - what is the scenario you think should be working but doesn't?

`wasm-bindgen` feature is an explicit opt-in on this crate and it's a dependency under `wasm32` target arch condition. There is neither a _reason_ or _a way_ to enable it on Linux targets. If you try to do that by accident, you'll see it fails as expected:

```
error[E0463]: can't find crate for `wasm_bindgen`
  --> src/lib.rs:41:1
   |
41 | extern crate wasm_bindgen;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
```

Reply:
Maybe if you can show a failing command, it will be easier to discuss, because right now I'm failing to see a scenario that doesn't work and accidentally compiles wrong target/feature combination.

Reply:
Unless you're arguing that it _should_ be possible to enable wasm-bindgen on non-Wasm targets and for it to behave as no-op? That's worse IMO than an explicit compiler error.

Reply:
Cargo has a global feature set per crate (as you also said yourself), so this way allows user who compiles a Wasm app with wasm-bindgen to set `wasm-bindgen` feature for `num_cpus` in their app, and have all the `num_cpus` instances return the Web value, while all other users just don't set this feature, and it will use the default implementation (whether libc for Linux/WASI/Emscripten/etc. or just `1` for others).

Reply:
> There is neither a reason or a way to enable it on Linux targets.

I believe we are talking past each other. It's not about you turning on the feature for Linux. It's about a dependency of yours rightfully activating the feature because it wants to know the number of cpus IF it gets compiled on WASM. However now the feature is always active FOR ALL TARGETS. So if you never even wanted to compile to WASM, the dependency in the middle will now make your code not compile to Linux anymore, the only target you actually wanted to compile for.

Reply:
> It's about a dependency of yours rightfully activating the feature

Well it shouldn't do that :D There is no reason for a library to enable this feature or toggle implementation details, it should be just using `num_cpus` normally and trust the value it returns.

This is very similar to, say, how you would use `rayon` on Wasm - you should just use rayon APIs and leave configuration of the global pool to the end app developer.

Unfortunately, I don't think there is a way to enforce such restrictions at Cargo level and prevent library developers from turning a feature on, but I think this kind of bugs will be quickly caught on practice.

Reply:
> There is no reason for a library to enable this feature

I don't understand this point. The library relies on num_cpus properly using wasm-bindgen, it needs to activate the feature. Though this is maybe a bit of a made up example, but in general libraries should activate the feature of their dependencies that they require. That's like the whole point of the feature resolving across the whole tree.

I mean I don't really care enough. This PR has a decently big chance to break my code, but at least with resolver 2 I have a decent chance to unbreak it. I'd at least advise to not land this before resolver 2 is on stable.

Reply:
> it needs to activate the feature

No, it shouldn't be activating this feature, just like it shouldn't be setting up a global rayon pool. This is not a feature for new APIs or anything like that, but an application-level config.

If you think many users might misuse it, I could as well remove the opt-in feature altogether, but that's worse IMO as it has a high chance of breaking non-wasm-bindgen wasm32 builds which suddenly will be relying on wasm-bindgen imports.

There is no option that is perfect, but I'm choosing to choose one where at least opt-in is very explicit and easy to catch locally or on CI upon the first run.

Reply:
I'm fairly sure that every crate tries to build themselves (either locally or on CI) on at least some non-Wasm target (Linux or Windows) and accidentally set `wasm-bindgen` feature would be immediately visible as a compilation failure, so the concern of such mistake making it all the way to crates.io in one of the popular dependencies seems very theoretical.

Reply:
Yeah I guess that's true at least. Though tbf I don't think people know that there even is such a thing as a feature that can't be used by a library crate. This is literally the first time I've ever heard of this in like close to 6 years of heavy Rust usage.

Reply:
Oh god, I just now see that you never even exposed this feature via cargo? Well then... that completely changes everything and dependencies can't even activate it. Then this is a sneaky way to get around this whole problem. Sorry for missing that this whole time.

Reply:
> Maybe you can add some documentation for this feature?

I agree, documenting that would be definitely useful. I'll add some wording tomorrow.

> Oh god, I just now see that you never even exposed this feature via cargo?

Hmm not sure what you mean by that. I mean, it's an optional dependency - so it still behaves like a feature and can be enabled by dependencies (I almost wish it didn't), but the code using it can be compiled only on `wasm32` target.

Reply:
> so it still behaves like a feature and can be enabled by dependencies

Yeah I thought you did some more complicated trickery than that for a short time (i.e. no feature at all and instead force the feature through rustc flags). But yeah instead this is simply just an optional crate that behaves like a normal cargo feature. So yeah this sorely needs documentation, most people wouldn't even see that this is meant to be activated from the outside if there's no features section in this Cargo.toml

I kind of agree by now that for this particular circumstance your solution makes sense. In general for most things having features that don't compile on every target is a huge no go, but here because of the way wasm-bindgen works (it's not even its own target) and wasm not even properly supporting threads, the binary needs an extended amount of control here and thus no library should even touch any of this at all. So that avoids the whole cargo feature resolver problem entirely.

Reply:
Thanks for your analysis, I definitely appreciate the conversation! Sounds like we've reached an agreement, and I'll update the docs tomorrow.

As I said, I'm not entirely happy with how this all works either, but, as you pointed out

> but here because of the way wasm-bindgen works (it's not even its own target) and wasm not even properly supporting threads

is the origin of this issue. I really wish wasm-bindgen just used a separate target, like Emscripten does, and not mixed with wasm32-unknown-unknown, but it's a bit too late for that.

Reply:
> I'll update the docs tomorrow

Went ahead and updated it now so that I don't forget :)

Reply:
Nice, I'll mark this as resolved then :)

Reply:
Weird. Let me mark it as such then.